### DISCUSSION TITLE
Implement waste-free SMC

### DISCUSSION THREAD

**Discussion by AdrienCorenflos** [blackjax-devs/blackjax#87] [New features]
A natural next step after the tempered SMC method is to implement the waste-free SMC algorithm in https://arxiv.org/abs/2011.02328. The gist is fairly simple: in order to work tempered SMC may require to run K >> 1 iterations of a chain for all N particles, and only keeps the last sample of each chain. This is quite statistically wasteful, and waste-free SMC proposes to reuse the samples generated along the chain to propose better estimates (in a sense this can be seen as the SMC analog of NUTS).

**Comment by rlouf** [comment]
This touches something we're already working on for HMC, so I'd wait until this is finished to work on waste-free SMC.

**Comment by AdrienCorenflos** [comment]
Seconded. This is opened as a reminder for the time being.

**Comment by junpenglao** [comment]
🤖 *This reply was drafted with Claude and reviewed by me (@junpenglao).*

Implemented in `blackjax/smc/waste_free.py`, usable as the resampling/MCMC strategy in tempered SMC. Closing.
